Rewrite of ProtocolConformance.md

mattrips opened the pull request:
Heavily re-written.  Please review fresh.  Pointless to review the changes.

mattrips committed:
Rewrite of ProtocolConformance.md

Heavily re-written.  Please review fresh.  Pointless to review the changes.

mattrips commented:
The heart of the document is Section 1.5. I believe that Section 1.5.4 captures the behavior related to generics and constrained implementations.  Please check me.

I have not yet addressed the behavior involving classed and implementations.

mattrips commented:
I believe the rules of specialization stated in section 1.5, could be more conveniently written as:

Given `S: P`, requirement *m* of `P`, and implementations *i<sub>1</sub>* and *i<sub>2</sub>* of *m*, the relative specialization between *i<sub>1</sub>* and *i<sub>2</sub>* is determined according to the following rules, which are stated in order of priority:

(i) if `S` is a concretization and the declaration of *i<sub>1</sub>* is more constrained than the declaration of `S: P` (as declared on generic `S`), *i<sub>1</sub>* is unavailable to serve as a witness for *m* and, so long as *i<sub>2</sub>* is not similarly unavailable, it is the more specialized;  

(ii) if *i<sub>1</sub>* is declared in an extension of a protocol and *i<sub>2</sub>* is declared on `S` (whether in its declaration and/or an extension), *i<sub>2</sub>* is the more specialized;

(iii) if *i<sub>1</sub>* is declared in an extension of `P` and *i<sub>2</sub>* is declared in an extension of a protocol that refines `P`, *i<sub>2</sub>* is the more specialized;

(iv) if *i<sub>1</sub>* and *i<sub>2</sub>* are both declared on the same protocol or the same concrete type and the declaration of *i<sub>1</sub>* is less constrained than the declaration of *i<sub>2</sub>*, *i<sub>2</sub>* is the more specialized; and

(v) otherwise, it is ambiguous which implementation is more specialized.

mattrips commented:
BTW, I dislike describing specialization as an _ad hoc_ set of rules.  Given the way the system operates, I don't see a way around that.  Do you see an alternative approach?

mattrips committed:
Fixed typos.

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
 # 1 Protocol Conformance
```
We need to choose a column limit.  I had wrapped at 80, I'd be happy with 100 if that works better for you, but we should agree on a number.

dabrahams replied:
You should integrate this material into the chapter list as part of the PR.

mattrips replied:
Thanks, Dave.  

I'm not used to writing prose with hard returns at the end of each line, so I've been a bit sloppy and undisciplined.  As I've been working on the Statements chapter, I've been getting much better at that.  Let's stick with 80.

Question about wrapping:  When the content ultimately is rendered for final use in a wider or narrower format than the selected , how does one deal with the out-of-place hard returns? 

Thank you for the substantive comments.  I will review them in detail, and revise accordingly.

dabrahams replied:
What editor are you using?  You might be able to set something up to wrap paragraphs when you save.

dabrahams replied:
The hard returns don't create a paragraph break unless you turn on [GFM](https://github.github.com/gfm/) in your Jekyll config

mattrips replied:
Thanks.  I'm using Sublime Text.  I've turned on the ruler, and set it at 80.  I think that will be enough to keep me on track.

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
 # 1 Protocol Conformance
+When a named concrete type[^1] `T` is conformed to a protocol `P`, a set of implementations--*witnesses*--is 
+determined, one for each requirement of `P`.   That set of witnesses is referred to as the protocol conformance
+of `T: P`.  This document specifies how Swift determines a conformance.
+[^1]: What about tuples and [SE-0283](https://github.com/apple/swift-evolution/blob/master/proposals/0283-tuples-are-equatable-comparable-hashable.md)?
```
There's a problem with the phrasing; there is no defined “when” for this.  We have three compilation phases to work with for talking about “when”: tokenization, parsing, and translation.

```suggestion
In a given scope, a concrete type `T` **conforms** to a protocol `P` if and only if:
- `T` is a nominal type and the conformance is implied [<-- link to section on how this implication is
  determined] by a
  *[type-inheritance-clause](https://docs.swift.org/swift-book/ReferenceManual/Types.html#grammar_type-inheritance-clause)* 
  visible in that scope.
- `T` is is a non-nominal type that implicitly conforms to `P` as described in [section dealing
  with tuples].

A protocol **conformance** of `T` to `P` is a triple (`T`, `P`, *M*), where *M* is a mapping from
each requirement *r* of `P` onto a corresponding **witness** that satisfies [<--link to requirement satisfaction] *r* for `T`.   

| Requirement | Witness |
|-|-|
| associated type | concrete type nested in `T` |
| non-`static` method | non-`static`, non-`class` method of `T` |
| non-operator `static` method | `static` or `class` method of `T`, or an `enum` case as described in [section describing that horrible new feature] |
| static operator | `static` or `class` operator of `T`, or a concrete module-scope operator |

This chapter describes how a conformance's mapping from requirements to witnesses is determined.
```

mattrips replied:
I like that approach.

For your consideration, I'd like to suggest:
- a `typealias` (declared or inferred) is the witness for an `associatedtype`
- for requirements, we should distinguish between `func` and `var` requirements
- we could use the term instance rather than non-`static`

dabrahams replied:
* typealiases really *are* aliases for types, rather than types themselves.  I think the witness is the actual type being aliased.  The typealias is just the thing that connects the witness to the conforming type.
* Ah yes, I left out properties.  And subscripts.  The witnesses are are actually getter functions or getter/setter pairs. I'lll write some text for that once we get this a little more settled.  We should drive this toward a state where it can be landed in master even if incomplete.
* “instance” works for me.
* I notice I also left out generic method /subscript requirements, whose instances are actually generics.

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
+protocol.  The syntactic declaration, `T: P`, that type `T` conforms to protocol `P` is separate and distinct 
+from the determination of the set of witnesses that is a conformance.  The substance of a conformance is not 
+declared.  It is created in response to a type being declared to conform to a protocol.  The set of witnesses 
+that is a conformance is inferred from the context.
```
Suggest dropping this paragraph.  The term was defined in my suggestion above.  There may not be any syntactic declaration:

```swift
struct X : Hashable { } // Implies Equatable
```

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
+declared.  It is created in response to a type being declared to conform to a protocol.  The set of witnesses 
+that is a conformance is inferred from the context.
+
+For a type `T` to conform to a protocol `P`, `T` must be declared to conform to `P`, and `T` must have at 
```
Not true for tuples

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
+that is a conformance is inferred from the context.
+
+For a type `T` to conform to a protocol `P`, `T` must be declared to conform to `P`, and `T` must have at 
+least one implementation for each protocol requirement of `P`.  A named concrete type is declared to conform to 
```
Well, "have" is poorly defined here.  The implementations may come from protocol extensions or be derived via associated type inference, etc.

Can we use the term “nominal type" please?  After all, this is named but not nominal:

```swift
typealias Named = (Int, Int)
```

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
+least one implementation for each protocol requirement of `P`.  A named concrete type is declared to conform to 
+a protocol in one of two ways.  A non-generic concrete type is expressly declared to conform to a protocol.  A 
+concretization of a generic type is implicitly declared to conform to a protocol based on the pattern established 
+by the generic type's declaration and extensions.
```
I think everything of value in this paragraph may be implied by my suggestion above?

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
+concretization of a generic type is implicitly declared to conform to a protocol based on the pattern established 
+by the generic type's declaration and extensions.
+
+><sub>***Conretization of a Generic Type**
```
Per the description at https://www.w3schools.com/tags/tag_sub.asp I can't understand why you'd want the subscript tag on this whole paragraph.

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
+
+><sub>***Conretization of a Generic Type**
+>A concretization of a generic type is a specialized version of the generic type formed by replacing the generic 
+ type's type parameters with concrete type arguments.  A generic type may specify that its concretizations will 
```
It's great that you wrote it down, but I think the definition of concretization probably belongs in https://dabrahams.github.io/SwiftRef/chapters/1600%20Generics.html

dabrahams replied:
We can't really talk about the generic type “specifying” anything.  The program may contain a  conformance declaration for a generic type.

I'm a little uncomfortable with the term “pattern” here, but I'm not entirely sure why so let it ride for now.

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
+ conform, or conditionally may conform, to one or more protocols, which establishes a pattern for protocol 
+ conformance.  When a concretization is formed, its protocol conformances are determined by that pattern.*</sub>
+
+A declaration that `T` conforms to `P` further constitutes, with respect to each protocol `O`*<sub>i</sub>* which 
```
```suggestion
A declaration that `T` conforms to `P` further constitutes, with respect to each protocol `O`*<sub>i</sub>* that 
```

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
+valid, the implicit declarations of each `T: O`*<sub>i</sub>* must be valid.  For each conformance `T: O`*<sub>i</sub>*, 
+a distinct set of protocol witnesses is established. 
+
+A type `T` cannot be declared to conform to a protocol `P` if, within the visible scope, another declaration 
+exists of `T: P`.  This rule holds true even where competing declarations are conditional with disjoint conditions.  
+In a given scope, a type can conform to a protocol in only one way.
```
I don't think that sentence adds anything.

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
+valid, the implicit declarations of each `T: O`*<sub>i</sub>* must be valid.  For each conformance `T: O`*<sub>i</sub>*, 
+a distinct set of protocol witnesses is established. 
+
+A type `T` cannot be declared to conform to a protocol `P` if, within the visible scope, another declaration 
+exists of `T: P`.  This rule holds true even where competing declarations are conditional with disjoint conditions.  
```
Well, that's only true if it's another *explicit* declaration of conformance.

The way to say this is:

```suggestion
If, in a given scope, more than one conformance declaration of a given type to a
given protocol is visible, the program is ill-formed.  This rule holds true even where
 competing declarations are conditional with disjoint conditions.  
```

Perhaps an example here.

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
+
+## 1.2 Requirements
+
+A **protocol requirement** *m* is a statement in the declaration of a protocol that a type declared to conform 
```
I think we need a top-level chapter called “Protocols” for this material.

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
+A **protocol requirement** *m* is a statement in the declaration of a protocol that a type declared to conform 
+to the protocol must have a member satisfying *m*.
+
+If one protocol refines another protocol, the requirements of the latter are not part of the requirements of 
+the former.  The requirements of a protocol are not overridden by a protocol that refines the protocol.  Thus, 
+the conformances formed by a refining protocol do not include witnesses for requirements of the refined protocol.
```
You said this 3 ways.  I'm not sure it *needs* to be said at all because it will be implied by a strict reading of the rest of the document, though I understand why you've included it: it was a point of confusion during discussions.  The right thing to do with stuff like that is to put it in a *note:*, which makes it “non-normative.”  I think it's important to separate the language *required* for understanding from the annotations and other supporting material.

dabrahams commented on ProtocolConformance.md:
```diff
@@ -1,146 +1,294 @@
+## 1.3 Implementations
+
+Given `T: P` and a protocol requirement *m* of `P`, an **implementation** of *m* is any member of `T` that satisfies 
+*m*.  A type may have more than one implementation of a requirement.
```
I don't think we want to define “implementation,” I think we should be talking about “witness candidates” for a given requirement.  When you call something an implementation of a requirement it implies some ongoing connection between the candidate and the requirement, but no such relationship exists.

It doesn't need to be a member, in the case of operators.

dabrahams reviewed (changes requested):
Hey Matt,

Great work on this.  I have lots of feedback and changes to suggest, but I want to stress how valuable it is an initial draft.  I think what I've said here may lead you to revisions of the second half, so I've paused halfway through to give you a chance to update it.

dabrahams commented:
> I believe the rules of specialization stated in section 1.5, could be more conveniently written as:

I think the rules are probably more complicated than that due to multiple parameters.  http://eel.is/c++draft/temp.func.order and http://eel.is/c++draft/temp.class.order provide a model for how to describe ordering that is almost certainly the basis for how it works in Swift.

mattrips commented:
> > I believe the rules of specialization stated in section 1.5, could be more conveniently written as:
> 
> I think the rules are probably more complicated than that due to multiple parameters.

When you say, "due to multiple parameters," to what are you referring?  In this context, aren't we dealing with identical function signatures, such that the parameters are the same for all candidates?

dabrahams commented:
On *ad hoc* rules: don't worry about it.  Every programming language has them, starting with syntax rules.

dabrahams commented:
Not sure; I'll have to look at this more closely.

On Mon, Jul 20, 2020 at 12:50 AM Matt Rips <notifications@github.com> wrote:

> I believe the rules of specialization stated in section 1.5, could be more
> conveniently written as:
>
> I think the rules are probably more complicated than that due to multiple
> parameters.
>
> When you say, "due to multiple parameters," to what are you referring? In
> this context, aren't we dealing with identical function signatures, such
> that the parameters are the same for all candidates?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dabrahams/SwiftRef/pull/2#issuecomment-660864466>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAKYIIZZDBXYKRFVAUOBQTR4PZKRANCNFSM4OQXHYSA>
> .
>


-- 
-Dave

mattrips committed:
Update ProtocolConformance.md with Suggestion from DA

Co-authored-by: Dave Abrahams <dabrahams@google.com>

mattrips committed:
Merge branch 'master' into mattrips-patch-1

mattrips committed:
Renamed ProtocolConformance.md to 1550 Protocols.md

mattrips committed:
Moved 1550 Protocols.md into chapters

mattrips committed:
Formatted 1550 Protocols.md in the style of a chapter

mattrips committed:
Added introduction and opening grammar to Protocols.md

mattrips committed:
Completed Outline of Protocols.md chapter

mattrips committed:
Inserted grammar of associatedtype and property for Protocols chapter.

mattrips commented:
@dabrahams, I'm revising the conformance writeup per your comments, and am building up the broader framework of the protocols chapter, to give it a good home.  For the initial pass, I'm keeping the chapter framework simple:  some grammar, a few introductory statements, and some stubs contemplating more details to be added in the future.  I should have it in reviewable form by tomorrow.

mattrips committed:
Completed the insertion of grammar into the protocol chapter.

mattrips committed:
Fleshed out the refinement section of the protocols chapter.

Much more to be added re: refinement, but this is a good start.

mattrips committed:
Completed Adoption and References sections of Protocol chapter.

mattrips committed:
Making progress on Conformance section of Protocol chapter.

And, removed .DS_store files.

dabrahams commented on chapters/1550 Protocols.md:
```diff
@@ -640,8 +640,40 @@ print(a.sorted()) // "(1, "one"), (2, "two")"
 ### Conformance
 ---
+
+A protocol **conformance** of type `T` to protocol `P` is a triple 
```
```suggestion
A (protocol) **conformance** of type `T` to protocol `P` is a triple 
```

mattrips committed:
Accept suggestion of parenthesizing protocol in (protocol) conformance

The intent being that, in the context of discussing protocols, it is distracting and unnecessary to prefix "conformance" with "protocol."  But, the parentheses also suggest that, outside that context, it often would be helpful to use the term "protocol conformance" when referencing the concept.

Co-authored-by: Dave Abrahams <dabrahams@google.com>

mattrips commented:
> a type “referenced by” another is not-a-thing. I believe the phrasing I suggested was perfectly correct. If you don't like “nested in” you could say “in the scope of”

Thank you.  The phrase "in the scope of" says it well.  To me, the "nested in" phrasing connotes the idea of a type declaration nested inside of another type declaration.

mattrips commented on chapters/1550 Protocols.md:
```diff
@@ -0,0 +1,1111 @@
+
+#### Method or Operator Function
+
+<div class="admonition grammar">
+  <p class="admonition-title">Grammar of a protocol method declaration</p>
+  <div class="syntax-group">
+    <p class="syntax-def">
+      <span class="name" id="protocol-method-declaration">
+      protocol-method-declaration</span>
+      <span class="arrow"> → </span> 
+      <a href="#function-head">function-head</a>
+      <a href="#function-name">function-name</a>
+      <a href="#generic-parameter-clause">
+      generic-parameter-clause</a><sub class="loosen">opt</sub>
+      <a href="#function-signature">function-signature</a>
+    </p>
+  </div>
+</div>
```
@dabrahams: Regarding usage of html rather than markup language for grammar. 

Reasons html tags are necessary:
1. To name syntax definitions for hyperlinking.
2. To get subscripting for "opt".

Reasons html tags are desirable:
1. To achieve pleasant formatting that makes for easier reading of a definition.

Reasons html tags are undesirable:
1. Visual clutter when editing.
1. Difficult to formulate from scratch.

Reasons why the visual clutter and difficult formulation are ok:
1. The html block is standardized, and easily cut, pasted and adapted.
1. The clutter isn't so bad when compared to the markup language version, which would look like:

```
GRAMMAR OF A PROTOCOL METHOD DECLARATION

<span class="name" id="protocol-method-declaration">
       protocol-method-declaration</span>
-> [function-head](#function-head) [function-name](#function-name) 
[generic-parameter-clause](#generic-parameter-clause)<sub>opt</sub>
[function-signature](#function-signature)
```

To my eye, the html tag version is easier to read and easier to write.

Thoughts?

mattrips committed:
Continuing to Draft Protocol Chapter

Filling in and Expanding Outline
Wrote Section for Property Requirement and Satisfaction Thereof
Added points re: Adoption by Generics
Building the Witness Section
Outline of Visibility Subsection
Returned to Drafting Conformance Section, and Rewrote with Better Formulation of the "More Specialized" Concept
Added Comments to Explain CSS
Added Temporary Links

mattrips committed:
Merge branch 'mattrips-patch-1' of https://github.com/dabrahams/SwiftRef into mattrips-patch-1

mattrips committed:
Reorder, and Clean Up in Protocol Chapter.

mattrips committed:
Revision of Conditions Applicable to Witness Eligibility in Protocol Chapter.

mattrips committed:
Reorganize Conformance Examples in Protocol Chapter.

mattrips committed:
Fleshed out protocol extension section of Protocol Chapter.

mattrips committed:
Coordination of Various Parts of Protocol Chapter, and Further Clean Up.

mattrips commented:
@dabrahams, I've pushed numerous commits related to the protocol chapter.  I think I've captured the spirit of your prior round of comments, but I very well may have missed a couple.  

Along the way, I've continued to flesh out the chapter.  Most of it is a work in progress, with many parts to be filled in and others to be added.  

You will be glad to see that I am (at last) religiously observing the 80 column mark, and, in many cases, am ending lines earlier so that the wraps occur in relation to clauses within a sentence.

Please note that I've heavily used the "adopt" term used in _The Swift Programming Language_, which is used there to connote the act of implicitly or expressly adopting a protocol.   I read "conform"  to mean something like the combination of adoption plus satisfaction of all requirements of a protocol.  I probably over did it, and there are cases where I've used "adopt" where it probably would be better to use "conform".

dabrahams commented:
Cool, will review your work as soon as I can.

“adopt” is a term inherited from Objective-C.  It means the same thing as
“conform.” You can't “adopt” a protocol without satisfying all its
requirements.  it's just exceedingly Apple-specific. No other language
community uses it that way.  We should use “conform” everywhere, IMO.

On Sun, Aug 2, 2020 at 3:30 PM Matt Rips <notifications@github.com> wrote:

> @dabrahams <https://github.com/dabrahams>, I've pushed numerous commits
> related to the protocol chapter. I think I've captured the spirit of your
> prior round of comments, but I very well may have missed a couple.
>
> Along the way, I've continued to flesh out the chapter. Most of it is a
> work in progress, with many parts to be filled in and others to be added.
>
> You will be glad to see that I am (at last) religiously observing the 80
> column mark, and, in many cases, am ending lines earlier so that the wraps
> occur in relation to clauses within a sentence.
>
> Please note that I've heavily used the "adopt" term used in *The Swift
> Programming Language*, which is used there to connote the act of
> implicitly or expressly adopting a protocol. I read "conform" to mean
> something like the combination of adoption plus satisfaction of all
> requirements of a protocol. I probably over did it, and there are cases
> where I've used "adopt" where it probably would be better to use "conform".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dabrahams/SwiftRef/pull/2#issuecomment-667732873>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAKYILFFG6RUVPVA3XHS2DR6XSHVANCNFSM4OQXHYSA>
> .
>


-- 
-Dave
